### ISSUE TITLE
Add support for compound indexes

### ISSUE THREAD

**Issue by diogotcorreia** [strapi/strapi#5972]
Related to #5682

I have the need to create some compound indexes (in my case, weighted text indexes). Also, indexes for timestamp fields would also be appreciated.

Anyway, I've been looking at `strapi-connector-mongoose` and the [`mount-models.js:280`](https://github.com/strapi/strapi/blob/master/packages/strapi-connector-mongoose/lib/mount-models.js#L280) file looks like the perfect place to insert the indexes.

This is what I have in mind (additional testing would need to be done):  
On `/api/**/models/**.settings.json`:
```jsonc
{
  "kind": "collectionType",
  "collection": "default",
  ...
  "indexes": [
    [
      // An array would allow for calling Schema.prototype.index() with options
      {
        "content": "text",
        "keywords": "text",
        "about": "text"
      },
      {
        "weights": {
          "content": 10,
          "keywords": 5
        },
        "name": "TextIndex"
      }
    ],
    { "someNumberField": -1, "anotherNumberField": 1 } // An object means no options
  ]
}
```

On `strapi-connector-mongoose/lib/mount-models.js`:
```javascript
if (_.isArray(definition.indexes)) {
  definition.indexes.forEach((index) => {
    if (_.isArray(index)) {
      schema.index(...index);
    } else {
      schema.index(index);
    }
  });
}
```

I'm just asking for a green light before submitting a PR.
Feel free to suggest any changes.

Useful links:
 - [MongoDB documentation on indexes](https://docs.mongodb.com/manual/indexes/#create-an-index)
 - [MongoDB documentation on text indexes](https://docs.mongodb.com/manual/core/index-text/)
 - [MongoDB documentation on weighted text indexes](https://docs.mongodb.com/manual/tutorial/control-results-of-text-search/)
 - [Mongoose documentation on schema indexes](https://mongoosejs.com/docs/api/schema.html#schema_Schema-index)

**Comment by nuno-tomas** [comment]
Looks good and would be an awesome improvement as we one can't live without compound indexes :)

**Comment by lauriejim** [comment]
Thank you for this proposal.
I ping @alexandrebodin about this.

**Comment by scisteffan** [comment]
Would this allow us to create a geospatial index?

**Comment by diogotcorreia** [comment]
> Would this allow us to create a geospatial index?

Yes, it would allow creating any type of index that you could create with `db.collection.createIndex`

**Comment by Convly** [comment]
Hi, thank you for this proposal.

It would indeed be awesome to bring compound indexes management to the model's schema.

I only have one concern about how to port this to other databases using bookshelf (`strapi-connector-bookshelf`), as I can't spot the related doc on https://bookshelfjs.org/api.html.

FYI, our goal is to have the same experience across all the databases. It means that in theory, the only modification you would need to change the DB is updating the connection name in the model's schema.

tl;dr: IMO, the idea is really cool, but maybe we should wait until we find a solution for bookshelf too, pinging @alexandrebodin as he might have a different opinion here.

**Comment by diogotcorreia** [comment]
@Convly 
Yeah, I agree with that. Although I don't use bookshelf myself, I think this is kind of a MongoDB specific feature. You can have indexes on bookshelf, but mostly for uniqueness, I guess.
Feedback on that is appreciated. I'll be sure to submit a PR once I get the green light from @alexandrebodin.

**Comment by nuno-tomas** [comment]
Hi @Convly,
I understand the need to keep db features kind of agnostic to the strapi ecosystem but would be great to atleast have some kind of workaround for this, even if it can evolve to something else in the future. 
Living without weighted text indexes (as an example) is a too big of a burden, in case you want some kind of smart/advance querying without having to add more tech debt. 
I ended up having to hack around to make it work, which doesn't make me fell either happy or proud :(

**Comment by derrickmehaffy** [comment]
Indeed, @Convly I also believe that where possible all databases should be equal in terms of features. I understand there may be some frustration when it comes to waiting for support on all databases in parity but Strapi is designed to be agnostic and as such the features should work equally (within reason) on all supported platforms (this is not just limited to Databases).

I've modified the title to represent this, but will also wait for @alexandrebodin's feedback as well.

**Comment by alexandrebodin** [comment]
Hi, IMO adding settings for mongo compound indexes is ok only if we can make the syntaxe work with the bookshelf connector, or at least partially and not breaking it. 

I would rather start by allowing to set those programmatically instead of forcing everything into settings. We could add in the Model.js file some callbacks that gets called with the schema so the user can do some manual changes like creating more complexe indexes

**Comment by derrickmehaffy** [comment]
> Hi, IMO adding settings for mongo compound indexes is ok only if we can make the syntaxe work with the bookshelf connector, or at least partially and not breaking it.
> 
> I would rather start by allowing to set those programmatically instead of forcing everything into settings. We could add in the Model.js file some callbacks that gets called with the schema so the user can do some manual changes like creating more complexe indexes

That sounds like a great idea @alexandrebodin

**Comment by diogotcorreia** [comment]
I'll be sure to open a PR later this week

**Comment by alexandrebodin** [comment]
@diogotcorreia Make sure to discuss what you want to implement before making a PR ;)

**Comment by diogotcorreia** [comment]
> @diogotcorreia Make sure to discuss what you want to implement before making a PR ;)

I'm thinking about implementing that callback in the model.js file you mentioned. Seems like a flexible approach that'll do much more than just add support for indexes.

**Comment by alexandrebodin** [comment]
@diogotcorreia Ok I think you should work on an rfcs for this so we make it work as a general concept for model lifecycles

**Comment by diogotcorreia** [comment]
I've just submitted a draft PR. Please take a look to see what you think.

**Comment by diogotcorreia** [comment]
As requested, I've opened an RFC for this. Please go and tell your opinion.

https://github.com/strapi/strapi/discussions/6681

**Comment by mattschlitz** [comment]
I may have missed this but is there some workaround for doing this in strapi as it exists today?

**Comment by diogotcorreia** [comment]
@mattschlitz Yes, there is, but it's not official. You have to edit the package inside `node_modules` and then use [`patch-package`](https://www.npmjs.com/package/patch-package) to make the changes persistent.
Take a look at my PR for guidance. #6515

**Comment by harrygusto2020** [comment]
Hi everyone, any status for this issue? I am using mongodb and become slowly without suitable indexing while the data growing up. Instead of edit the node_modules, any other workaround for doing this?

**Comment by derrickmehaffy** [comment]
@harrygusto2020 currently no, there is no updates on this feature request.

**Comment by IPWright83** [comment]
I'd be interested to get an update on this one too (or at least finding a way to unblock it, looks like it's been stuck for circa 4 months?) - I've built some stuff assuming I could just add a custom index to MongoDB (not realising Strapi would then choose to delete it) which is quite restricting. 

@diogotcorreia looks like he's put in quite some effort and tried to address the bookshelf concerns too in https://github.com/strapi/strapi/pull/6515. Is there something still missing (it's a little unclear to me from the threads) as to the problems with this PR? I'm trying to keep my discussion here as per @alexandrebodin 's comment, but can move to the PR if preferred.

**Comment by harrygusto2020** [comment]
Agreed, for me this is a critical issue about performance, MongoDB must perform a collection scan for every document if without indexes. I think we should find a workaround way to solve it at least.

**Comment by derrickmehaffy** [comment]
Related thread for MongoDB users: https://forum.strapi.io/t/best-approaching-for-adding-custom-indexes

**Comment by trongrg** [comment]
I ended up using the patch-package workaround, but it would be nice to have this feature built-in

Anything could be done to make this in the radar again?

**Comment by darron1217** [comment]
@diogotcorreia Thank you for your great work 👍 
Anyway, I guess the 'rfc' which @alexandrebodin said is on`strapi/rfcs` repo.
https://github.com/strapi/rfcs

Would you submit rfc there?

**Comment by diogotcorreia** [comment]
@darron1217 
It's already there https://github.com/strapi/rfcs/pull/15

**Comment by darron1217** [comment]
@diogotcorreia Oh, I was stupid... haha
Thanks again for your contribution :)

**Comment by derrickmehaffy** [comment]
Marking as closed as we will be dropping MongoDB support in the v4: https://forum.strapi.io/t/dropping-mongodb-support/4549
